kratos/mpi/includes/mpi_communicator.h

```diff
@@ -74,12 +67,502 @@ namespace Kratos
+
+    using ValueType = Kratos::Flags;
+    using SendType = typename SendTraits<ValueType>::SendType;
+
+    NodalFlagsAccess(const Kratos::Flags& rMask):
+        NodalContainerAccess(),
+        mrMask(rMask)
+    {}
+
+    Kratos::Flags& GetValue(IteratorType& iter)
+    {
+        return *iter;
+    }
+
+    const Kratos::Flags& GetValue(const ConstIteratorType& iter)
+    {
+        return *iter;
+    }
+};
+
+class NodalSolutionStepDataAccess: public NodalContainerAccess
+{
+public:
+
+    using ValueType = Kratos::VariablesListDataValueContainer;
+    using SendType = typename SendTraits<ValueType>::SendType;
+
+    ValueType& GetValue(IteratorType& iter)
+    {
+        return iter->SolutionStepData();
+    }
+
+    const ValueType& GetValue(const ConstIteratorType& iter)
+    {
+        return iter->SolutionStepData();
+    }
+};
+
+class DofIdAccess: public NodalContainerAccess
+{
+public:
+
+    using ValueType = Node<3>::DofsContainerType;
+    using SendType = typename SendTraits<ValueType>::SendType;
+
+    ValueType& GetValue(IteratorType& iter)
+    {
+        return iter->GetDofs();
+    }
 
-/** Detail class definition.
- */
+    const ValueType& GetValue(const ConstIteratorType& iter)
+    {
+        return iter->GetDofs();
+    }
+};
+
+template<class TValue> class ElementalDataAccess: public ElementalContainerAccess
+{
+    const Variable<TValue>& mrVariable;
+
+public:
+
+    using ValueType = TValue;
+    using SendType = typename SendTraits<TValue>::SendType;
+
+    ElementalDataAccess(const Variable<TValue>& mrVariable):
+        ElementalContainerAccess(),
+        mrVariable(mrVariable)
+    {}
+
+    TValue& GetValue(IteratorType& iter)
+    {
+        return iter->GetValue(mrVariable);
+    }
+
+    const TValue& GetValue(const ConstIteratorType& iter)
+    {
+        return iter->GetValue(mrVariable);
+    }
+};
+
+class ElementalFlagsAccess: public ElementalContainerAccess
+{
+public:
+
+    const Kratos::Flags& mrMask;
+
+    using ValueType = Kratos::Flags;
+    using SendType = typename SendTraits<ValueType>::SendType;
+
+    ElementalFlagsAccess(const Kratos::Flags& rMask):
+        ElementalContainerAccess(),
+        mrMask(rMask)
+    {}
+
+    Kratos::Flags& GetValue(IteratorType& iter)
+    {
+        return *iter;
+    }
+
+    const Kratos::Flags& GetValue(const ConstIteratorType& iter)
+    {
+        return *iter;
+    }
+};
+
+}
+
+/// MPICommunicator manages the transfer of ModelPart data in MPI distributed memory environment.
 class MPICommunicator : public Communicator
 {
+
+enum class DistributedType {
+    Local,
+    Ghost
+};
+
+// Auxiliary type for compile-time dispatch of local/ghost mesh access
+template<DistributedType TDistributed> struct MeshAccess
+{
+    constexpr static DistributedType Value = TDistributed;
+};
+
+enum class OperationType {
+    Replace,
+    SumValues,
+    MinValues,
+    OrMaskedFlags,
+    AndMaskedFlags,
+    ReplaceMaskedFlags
+};
```

Comment:
Just a minor comment, Defined wouldn't be more descriptive than Masked?

Reply:
I don't love either name: the idea is that this "operation" works only on the flags passed as argument to the AccessType:
```
UpdateValues(..., NodalFlagAccess(BOUNDARY), OrMaskedFlags);
```
Modifies only the `BOUNDARY` flag, regardless of which nodal flags are defined. Personally, if it was called "defined", I would understand it to affect all defined flags in the node.

Reply:
Maybe `OrSelectedFlags`? Suggestions are welcome...

Reply:
Now I see the meaning, saying that the name was not descriptive! 

I like `OrSelectedFlags` more.

Reply:
`OrGivenFlags`?

Reply:
I went ahead with `OrAccessedFlags` to try to convey that it affects the flags given by the "Access" object:
```
        constexpr MeshAccess<DistributedType::Local> local_meshes;
        constexpr MeshAccess<DistributedType::Ghost> ghost_meshes;
        MPIInternals::NodalFlagsAccess nodal_flag_access(TheFlags);
        constexpr Operation<OperationType::OrAccessedFlags> or_accessed;
        constexpr Operation<OperationType::ReplaceAccessedFlags> replace_accessed;

        TransferDistributedValues(ghost_meshes, local_meshes, nodal_flag_access, or_accessed);
```
I'm open to change if someone has a better name.

Reply:
:+1: